hugr-core/src/std_extensions/collections/static_array.rs

```diff
@@ -0,0 +1,396 @@
+/// Statically sized array of values, all of the same type.
+pub struct StaticArrayValue {
+    /// The contents of the `StaticArrayValue`.
+    pub value: ArrayValue,
+    /// The name of the `StaticArrayValue`.
+    pub name: String,
+}
+
+impl StaticArrayValue {
+    delegate! {
+        to self.value {
+            /// Returns the type of values inside the `[StaticArrayValue]`.
+            pub fn get_element_type(&self) -> &Type;
+        }
+
+    }
+
+    /// Create a new [CustomConst] for an array of values of type `typ`.
+    /// That all values are of type `typ` is not checked here.
+    pub fn new(name: impl ToString, typ: Type, contents: impl IntoIterator<Item = Value>) -> Self {
+        Self {
+            value: ArrayValue::new(typ, contents),
+            name: name.to_string(),
+        }
+    }
+
+    /// Create a new [CustomConst] for an empty array of values of type `typ`.
+    pub fn new_empty(name: impl ToString, typ: Type) -> Self {
+        Self::new(name, typ, iter::empty())
+    }
+
+    /// Returns the type of the `[StaticArrayValue]` as a `[CustomType]`.`
+    pub fn custom_type(&self) -> CustomType {
+        static_array_custom_type(self.get_element_type().clone())
+    }
+}
+
+impl TryHash for StaticArrayValue {
+    fn try_hash(&self, st: &mut dyn hash::Hasher) -> bool {
+        st.write(self.name.as_bytes());
+        self.value.try_hash(st)
+    }
+}
+
+#[typetag::serde]
+impl CustomConst for StaticArrayValue {
+    fn name(&self) -> ValueName {
+        ValueName::new_inline("const_array")
+    }
+
+    fn get_type(&self) -> Type {
+        self.custom_type().into()
+    }
+
+    fn equal_consts(&self, other: &dyn CustomConst) -> bool {
+        crate::ops::constant::downcast_equal_consts(self, other)
+    }
+
+    delegate! {
+        to self.value {
+            fn update_extensions(
+                &mut self,
+                extensions: &WeakExtensionRegistry,
+            ) -> Result<(), ExtensionResolutionError>;
+
+            fn extension_reqs(&self) -> ExtensionSet;
+        }
+    }
+}
+
+lazy_static! {
+    /// Extension for array operations.
+    pub static ref EXTENSION: Arc<Extension> = {
+        use TypeBound::Copyable;
+        Extension::new_arc(EXTENSION_ID.to_owned(), VERSION, |extension, extension_ref| {
+            extension.add_type(
+                    STATIC_ARRAY_TYPENAME,
+                    vec![Copyable.into()],
+                    "Fixed-length constant array".into(),
+                    Copyable.into(),
+                    extension_ref,
+                )
+                .unwrap();
+
+            StaticArrayOpDef::load_all_ops(extension, extension_ref).unwrap();
+        })
+    };
+}
+
+fn instantiate_const_static_array_custom_type(
+    def: &TypeDef,
+    element_ty: impl Into<TypeArg>,
+) -> CustomType {
+    def.instantiate([element_ty.into()])
+        .expect("const array parameters are valid")
```

Comment:
This `expect` will fail if someone passes a non-copyable type. Maybe a more informative panic message would be good?